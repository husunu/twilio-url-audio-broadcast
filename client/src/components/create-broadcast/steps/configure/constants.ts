export const buyNewNumberUrl =
  'https://www.twilio.com/console/phone-numbers/search';
