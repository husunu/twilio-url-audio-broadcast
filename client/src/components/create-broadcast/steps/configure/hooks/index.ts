export { useConfigureState } from './useConfigureState';
export { useNumbers } from './useNumbers';
export { useTestCall } from './useTestCall';
