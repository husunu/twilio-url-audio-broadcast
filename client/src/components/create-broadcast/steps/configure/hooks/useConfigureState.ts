import { useCallback, useEffect, useMemo, useState } from 'react';
import isEmpty from 'lodash.isempty';
import { ConfigureStep } from '@/types';
import { AlertDto } from '@/components/common/alert';

export const useConfigureState = (data: ConfigureStep) => {
  const [from, setFrom] = useState(data.number);
  const [message, setMessage] = useState(data.message);
  const [name, setName] = useState(data.name);
  const [testNumber, setTestNumber] = useState<string>('');
  const [alert, setAlert] = useState<AlertDto>({});

  useEffect(() => {
    setFrom(data.number);
    setMessage(data.message);
    setName(data.name);
  }, [data]);

  const isValid = useMemo(() => {
    return !isEmpty(name) && !isEmpty(message) && !isEmpty(from);
  }, [from, message, name]);

  const resetAlert = useCallback(() => {
    setAlert({});
  }, [setAlert]);

  return {
    alert,
    setAlert,
    resetAlert,
    isValid,
    from,
    setFrom,
    message,
    setMessage,
    name,
    setName,
    testNumber,
    setTestNumber,
  };
};
