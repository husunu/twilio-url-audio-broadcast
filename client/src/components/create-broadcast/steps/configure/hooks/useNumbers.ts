import { useCallback, useEffect, useState } from 'react';
import useFetch from 'use-http';
import { IncomingPhoneNumberInstance } from 'twilio/lib/rest/api/v2010/account/incomingPhoneNumber';

export const useNumbers = () => {
  const [numbers, setNumbers] = useState<IncomingPhoneNumberInstance[]>([]);
  const { request } = useFetch<IncomingPhoneNumberInstance[]>();

  const loadNumbers = useCallback(() => {
    (async () => {
      const response = await request.get('/numbers');
      setNumbers(response);
    })();
  }, [setNumbers, request]);

  useEffect(() => {
    loadNumbers();
  }, [request, loadNumbers]);
  return { numbers, loadNumbers };
};
