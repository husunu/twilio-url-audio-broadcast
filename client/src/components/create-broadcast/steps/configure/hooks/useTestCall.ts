import { useCallback, useState, useMemo } from 'react';
import isEmpty from 'lodash.isempty';
import useFetch from 'use-http';
import { CallStatus } from '@/types';
import socketIOClient from 'socket.io-client';

const wasCallSuccesful = (callStatus: CallStatus) => {
  return [
    CallStatus.COMPLETED,
    CallStatus.BUSY,
    CallStatus.NO_ANSWER,
    CallStatus.CANCELED,
  ].includes(callStatus);
};

export const useTestCall = (from: string, to: string, message: string) => {
  const [request, response] = useFetch();
  const [cancelRequest] = useFetch();
  const [callStatus, setCallStatus] = useState('initiating');
  const [isTestCallOngoing, setIsTestCallOngoing] = useState(false);
  const [callSid, setCallSid] = useState('');
  const isCancelLoading = useMemo(() => cancelRequest.loading, [cancelRequest]);
  const isCancelAvailable = useMemo(() => !isEmpty(callSid), [callSid]);

  const cancelTestCall = useCallback(async () => {
    await cancelRequest.get(`/test-call/cancel/${callSid}`);
  }, [callSid, cancelRequest]);

  const makeTestCall = useCallback(() => {
    return new Promise(async (resolve, reject) => {
      try {
        setCallStatus('initiating');
        setIsTestCallOngoing(true);

        console.log(
          'making a test call to: %s, from: %s with the message: %s',
          to,
          from,
          message
        );

        await request.post('/test-call', {
          message,
          to,
          from,
        });

        if (!response.ok) {
          setIsTestCallOngoing(false);
          setCallStatus(CallStatus.FAILED);
          reject();
        }

        console.log('response %i', response.status);
        console.log('callSid %s', response.data?.callSid);
        setCallSid(response.data?.callSid);

        const socket = socketIOClient(window.location.origin);

        socket.on('update', (value: any) => {
          console.log('document updated: ', value);
          console.log('status: ', value.status);

          setCallStatus(value.status);

          if (wasCallSuccesful(value.status)) {
            setCallSid('');
            setIsTestCallOngoing(false);
            socket.disconnect();
            resolve(value.status);
          }

          if (value.status === CallStatus.FAILED) {
            setIsTestCallOngoing(false);
            socket.disconnect();
            reject();
          }
        });
      } catch (error) {
        console.log('second error', error);
        reject(error);
      }
    });
  }, [to, from, message, request, response.ok, response.status, response.data]);

  return {
    makeTestCall,
    cancelTestCall,
    isCancelLoading,
    isCancelAvailable,
    callStatus,
    isTestCallOngoing,
  };
};
