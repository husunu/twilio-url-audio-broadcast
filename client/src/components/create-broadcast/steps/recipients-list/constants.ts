export const MAX_NUMBERS = 500;

export const ALERTS = {
  FILE_READ_ERROR: 'There was an error reading the file.',
  FILE_EXCEEDS_MAX: `File has more than ${MAX_NUMBERS} numbers. Please review issues and upload again.`,
  INVALID_NUMBERS:
    'The list of recipients has invalid numbers. Please review issues and upload again.',
  PARSE_SUCCESS: 'The list of recipients is successfully loaded.',
  FILE_REJECT_ERROR: 'Please make sure you tried to upload a text file (.txt)',
};

export const INVALID_NUMBERS = 'invalidNumbers';
export const VALID_NUMBERS = 'validNumbers';

export const INITIAL_VALIDATION_RESULT = {
  [INVALID_NUMBERS]: [],
  [VALID_NUMBERS]: [],
};
