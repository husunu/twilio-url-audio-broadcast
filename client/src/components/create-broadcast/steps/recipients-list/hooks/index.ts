export { useCustomDropZone } from './useCustomDropZone';
export { useFileReader } from './useFileReader';
export { useNumbersValidator } from './useNumbersValidator';
export { useAlertSetter } from './useAlertSetter';
export { useRecipientsList } from './useRecipientsList';
