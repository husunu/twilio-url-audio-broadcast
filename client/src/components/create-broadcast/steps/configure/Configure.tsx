import React, { ChangeEvent, useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { isValidNumber } from 'libphonenumber-js/';
import isEmpty from 'lodash.isempty';
import { completeConfigureStep } from '@/redux';
import { ConfigureStep } from '@/types';
import { Alert, AlertType } from '@/components/common/alert';
import { buyNewNumberUrl } from './constants';
import { useNumbers, useTestCall, useConfigureState } from './hooks';

import {
  BaseNavigation,
  Wizard,
  StepProps,
} from '@/components/create-broadcast';

import {
  TextField,
  FormControl,
  InputLabel,
  MenuItem,
  Select,
  Button,
  Grid,
  Divider,
  Box,
  CircularProgress,
  Typography,
} from '@material-ui/core';

import RefreshIcon from '@material-ui/icons/Refresh';
import useStyles, { WhiteTextField } from './styles';

export const Configure = ({ step, data }: StepProps) => {
  const dispatch = useDispatch();
  const classes = useStyles();
  const { numbers, loadNumbers } = useNumbers();

  const {
    isValid,
    from = '',
    setFrom,
    message = '',
    setMessage,
    name = '',
    setName,
    testNumber = '',
    setTestNumber,
    alert,
    setAlert,
    resetAlert,
  } = useConfigureState(data as ConfigureStep);

  const {
    makeTestCall,
    cancelTestCall,
    isCancelLoading,
    isCancelAvailable,
    callStatus,
    isTestCallOngoing,
  } = useTestCall(from, testNumber, message);

  const onComplete = useCallback(() => {
    dispatch(
      completeConfigureStep({
        number: from,
        message,
        name,
        completed: true,
      })
    );
  }, [from, message, name, dispatch]);

  const handleNumberSelectChange = useCallback(
    ({ target: { value } }: ChangeEvent<any>) => {
      if (value === 'new') {
        window.open(buyNewNumberUrl, '_blank');
      } else {
        setFrom(value);
      }
    },
    [setFrom]
  );

  const testTheNumberHandler = useCallback(async () => {
    resetAlert();
    try {
      const finalStatus = await makeTestCall();
      setAlert({
        message: `Test call was ${finalStatus}`,
        type: AlertType.INFO,
      });
    } catch (error) {
      setAlert({
        message: 'It was not possible to complete the call.',
        type: AlertType.ERROR,
      });
    }
  }, [makeTestCall, setAlert, resetAlert]);

  const isTestCallButtonDisabled = useMemo(
    () =>
      !isValidNumber(testNumber) ||
      isTestCallOngoing ||
      isEmpty(message) ||
      isEmpty(testNumber) ||
      isEmpty(from),
    [from, isTestCallOngoing, message, testNumber]
  );

  return (
    <>
      <Wizard.Step step={step}>
        <Typography variant="h4" component="h2">
          1. Configure
        </Typography>
        <Box mt={1}>
          <Typography>
            Enter your voice notifications details and test your notification
          </Typography>
        </Box>
        <TextField
          id="name"
          inputProps={{
            'data-testid': 'broadcast-name',
          }}
          value={name}
          fullWidth
          label="Notification name"
          placeholder="Notification name is required"
          variant="outlined"
          margin="normal"
          onChange={(e) => setName(e.target.value.substr(0, 250))}
        />
        <Grid container alignItems="center" spacing={2}>
          <Grid item xs={6}>
            <FormControl fullWidth margin="normal" variant="outlined">
              <InputLabel id="from-label">From Number</InputLabel>
              <Select
                id="from"
                labelId="from-label"
                value={from}
                label="Choose your number"
                onChange={handleNumberSelectChange}
                inputProps={{
                  'data-testid': 'select-id',
                }}
              >
                {numbers &&
                  numbers.map(({ phoneNumber }) => (
                    <MenuItem key={phoneNumber} value={phoneNumber}>
                      {phoneNumber}
                    </MenuItem>
                  ))}
                <Divider />
                <MenuItem key="new" value="new">
                  Buy Twilio phone number
                </MenuItem>
              </Select>
            </FormControl>
          </Grid>
          <Grid item xs={6}>
            <Button
              variant="outlined"
              color="secondary"
              size="large"
              style={{ top: '3px' }}
              onClick={loadNumbers}
              startIcon={<RefreshIcon />}
            >
              Refresh number list
            </Button>
          </Grid>
        </Grid>
        <TextField
          id="message"
          fullWidth
          label="Audio url"
          placeholder="Audio url"
          variant="outlined"
          margin="normal"
          multiline
          value={message}
          rows={1}
          onChange={(e) => setMessage(e.target.value.substr(0, 3000))}
          helperText="Example: https://example.com/audio.mp3"
        />
        <Grid container className={classes.testContainer}>
          <Grid item xs={12}>
            <Typography variant="h5" component="h2">
              Test your voice notification (optional)
            </Typography>
            <Typography>
              Enter your phone number to receive a call to test your voice
              notification before sending. Please note that this test could take
              up to a few minutes to complete.
            </Typography>
          </Grid>
          <Grid container alignItems="center" spacing={2}>
            <Grid item xs={6}>
              <WhiteTextField
                id="test-number"
                fullWidth
                label="Phone number"
                placeholder="Enter broadcast to send"
                variant="outlined"
                margin="normal"
                helperText="Phone number should be in international format and start with +XX code"
                value={testNumber}
                onChange={(e) => setTestNumber(e.target.value)}
                disabled={isTestCallOngoing}
                data-testid="test-from-input-number"
              />
            </Grid>
            <Grid item xs={6}>
              <Button
                variant="outlined"
                color="primary"
                style={{ top: '-5px' }}
                disabled={isTestCallButtonDisabled}
                size="large"
                onClick={testTheNumberHandler}
                data-testid="test-button"
              >
                Send Test
              </Button>
            </Grid>
            <Grid item xs={12}>
              <Alert alert={alert} />
              {isTestCallOngoing && (
                <Box
                  display="flex"
                  justifyContent="center"
                  alignItems="center"
                  flexDirection="column"
                >
                  <Typography>Test call status: {callStatus}</Typography>
                  <CircularProgress style={{ margin: '5px 0' }} />
                  <Button
                    disabled={isCancelLoading || !isCancelAvailable}
                    onClick={cancelTestCall}
                  >
                    Cancel
                  </Button>
                </Box>
              )}
            </Grid>
          </Grid>
        </Grid>
      </Wizard.Step>
      <BaseNavigation
        currentStep={step}
        onComplete={onComplete}
        canGoNext={isValid && !isTestCallOngoing}
      />
    </>
  );
};
