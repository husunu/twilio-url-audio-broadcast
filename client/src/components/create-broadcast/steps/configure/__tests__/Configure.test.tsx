import React from 'react';
import { FetchMock } from 'jest-fetch-mock/types';
import { renderWithAppContexts } from '@/testUtils';
import { StepProps } from '@/components/create-broadcast/types';
import { steps } from '../../../constants';
import { Configure } from '../Configure';
import { useTestCall } from '../hooks/useTestCall';
import { CallStatus } from '@/types';
import UserEvent from '@testing-library/user-event';

import {
  waitFor,
  cleanup,
  fireEvent,
  RenderResult,
  within,
} from '@testing-library/react';

jest.mock('../hooks/useTestCall');

export const selectMaterialUiSelectOption = (
  element: any,
  optionText: string
) => {
  // The the button that opens the dropdown, which is a sibling of the input
  const selectButton = element.parentNode.querySelector('[role=button]');

  // Open the select dropdown
  UserEvent.click(selectButton);

  // Get the dropdown element. We don't use getByRole() because it includes <select>s too.
  const listbox = document.body.querySelector<HTMLUListElement>(
    'ul[role=listbox]'
  );

  // Click the list item
  const listItem = within(listbox!).getByText(optionText);
  UserEvent.click(listItem);
};

const fetch = global.fetch as FetchMock;
const testNumber = '+34555666777';
const useTestCallMock = useTestCall as jest.MockedFunction<typeof useTestCall>;

describe('Configure', () => {
  let wrapper: RenderResult;
  const step = steps[0];
  const props: StepProps = {
    data: {
      completed: false,
      name: '',
      number: '',
      message: 'Ahoy test!',
    },
    step,
  };

  beforeEach(async () => {
    fetch.mockResponse(JSON.stringify([{ phoneNumber: testNumber }]));
  });

  afterEach(() => {
    cleanup();
    fetch.resetMocks();
  });

  test('renders the page', async () => {
    useTestCallMock.mockReturnValue({
      makeTestCall: jest.fn(),
      callStatus: CallStatus.INITIATED,
      isTestCallOngoing: false,
      cancelTestCall: jest.fn(),
      isCancelAvailable: false,
      isCancelLoading: false,
    });

    wrapper = renderWithAppContexts(<Configure {...props} />);
    expect(wrapper.container).toMatchSnapshot();

    await waitFor(() => {
      expect(fetch.mock.calls.length).toBe(1);
    });
  });

  test('should have disabled the test btn when data is missing', async () => {
    wrapper = renderWithAppContexts(<Configure {...props} />);

    await waitFor(() => {
      expect(fetch.mock.calls.length).toBe(1);
    });

    const { getByTestId } = wrapper;
    expect(getByTestId('test-button')).toBeDisabled();

    const selectNode = getByTestId('select-id') as HTMLInputElement;
    selectMaterialUiSelectOption(selectNode, testNumber);

    const fromInput = getByTestId('test-from-input-number').querySelector(
      'input'
    )!;

    fireEvent.change(fromInput, { target: { value: '+34626222111' } });
    expect(getByTestId('test-button')).not.toBeDisabled();
  });

  test('click test should make a call', async () => {
    const makeTestCall = jest.fn();

    useTestCallMock.mockImplementation(() => ({
      makeTestCall,
      callStatus: 'initiating',
      isTestCallOngoing: false,
      cancelTestCall: jest.fn(),
      isCancelAvailable: false,
      isCancelLoading: false,
    }));

    wrapper = renderWithAppContexts(<Configure {...props} />);

    await waitFor(() => {
      expect(fetch.mock.calls.length).toBe(1);
    });

    const { getByTestId } = wrapper;
    selectMaterialUiSelectOption(getByTestId('select-id'), testNumber);

    fireEvent.change(
      getByTestId('test-from-input-number').querySelector('input')!,
      { target: { value: '+34626222111' } }
    );

    fireEvent.click(getByTestId('test-button'));

    await waitFor(() => {
      expect(makeTestCall).toHaveBeenCalled();
    });
  });

  test('when reciving a new status, should be updated in the UI', async () => {
    const makeTestCall = jest.fn();

    useTestCallMock.mockImplementation(() => ({
      makeTestCall,
      callStatus: CallStatus.IN_PROGRESS,
      isTestCallOngoing: true,
      cancelTestCall: jest.fn(),
      isCancelAvailable: false,
      isCancelLoading: false,
    }));

    wrapper = renderWithAppContexts(<Configure {...props} />);
    const { getByText } = wrapper;

    expect(getByText(/in-progress/i)).toBeInTheDocument();

    await waitFor(() => {
      expect(fetch.mock.calls.length).toBe(1);
    });
  });

  test('when a test-call is ongoing the cancel button should be enabled', async () => {
    useTestCallMock.mockImplementation(() => ({
      makeTestCall: jest.fn(),
      callStatus: CallStatus.IN_PROGRESS,
      isTestCallOngoing: true,
      cancelTestCall: jest.fn(),
      isCancelAvailable: true,
      isCancelLoading: false,
    }));

    wrapper = renderWithAppContexts(<Configure {...props} />);
    const { getByText } = wrapper;

    await waitFor(() => {
      expect(getByText(/cancel/i)).toBeInTheDocument();
      expect(getByText(/cancel/i)).not.toBeDisabled();
    });
  });

  test('when cancel test is loading the test button should be disabled', async () => {
    useTestCallMock.mockImplementation(() => ({
      makeTestCall: jest.fn(),
      callStatus: CallStatus.IN_PROGRESS,
      isTestCallOngoing: true,
      cancelTestCall: jest.fn(),
      isCancelAvailable: true,
      isCancelLoading: true,
    }));

    wrapper = renderWithAppContexts(<Configure {...props} />);

    const { getByText } = wrapper;

    await waitFor(() => {
      expect(getByText(/cancel/i)).toBeInTheDocument();
      expect(getByText(/cancel/i)).toBeDisabled();
    });
  });

  test('when the call is not yet processed the cancel button should be disabled', async () => {
    useTestCallMock.mockImplementation(() => ({
      makeTestCall: jest.fn(),
      callStatus: CallStatus.IN_PROGRESS,
      isTestCallOngoing: true,
      cancelTestCall: jest.fn(),
      isCancelAvailable: false,
      isCancelLoading: false,
    }));

    wrapper = renderWithAppContexts(<Configure {...props} />);
    const { getByText } = wrapper;

    await waitFor(() => {
      expect(getByText(/cancel/i)).toBeInTheDocument();
      expect(getByText(/cancel/i)).toBeDisabled();
    });
  });
});
