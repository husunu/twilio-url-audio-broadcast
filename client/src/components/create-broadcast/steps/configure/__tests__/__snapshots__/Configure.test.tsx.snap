// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure renders the page 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root-2 WithStyles(ForwardRef(Card))-root-3 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-34"
    >
      <h2
        class="MuiTypography-root MuiTypography-h4"
      >
        1. Configure
      </h2>
      <div
        class="MuiBox-root MuiBox-root-65"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Enter your voice notifications details and test your notification
        </p>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="name"
          id="name-label"
        >
          Notification name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="broadcast-name"
            id="name"
            placeholder="Notification name is required"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-126"
            >
              <span>
                Notification name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              id="from-label"
            >
              From Number
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="from-label from"
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                id="from"
                role="button"
                tabindex="0"
              >
                <span>
                  ​
                </span>
              </div>
              <input
                data-testid="select-id"
                type="hidden"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-126"
                >
                  <span>
                    Choose your number
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
            style="top: 3px;"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeLarge"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
                  />
                </svg>
              </span>
              Refresh number list
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink="true"
          for="message"
          id="message-label"
        >
          Message
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
        >
          <textarea
            aria-describedby="message-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
            id="message"
            placeholder="Message"
            rows="4"
          >
            Ahoy test!
          </textarea>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-126 PrivateNotchedOutline-legendNotched-127"
            >
              <span>
                Message
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
          id="message-helper-text"
        >
          Message can’t exceed 3,000 characters
        </p>
      </div>
      <div
        class="MuiGrid-root makeStyles-testContainer-1 MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5"
          >
            Test your voice notification (optional)
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Enter your phone number to receive a call to test your voice notification before sending. Please note that this test could take up to a few minutes to complete.
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-294 MuiFormControl-marginNormal MuiFormControl-fullWidth"
              data-testid="test-from-input-number"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
                for="test-number"
                id="test-number-label"
              >
                Phone number
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-describedby="test-number-helper-text"
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="test-number"
                  placeholder="Enter broadcast to send"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-126"
                  >
                    <span>
                      Phone number
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-contained"
                id="test-number-helper-text"
              >
                Phone number should be in international format and start with +XX code
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge Mui-disabled Mui-disabled"
              data-testid="test-button"
              disabled=""
              style="top: -5px;"
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Send Test
              </span>
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-295"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Discard
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
        style="text-align: right;"
      >
        
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
          data-testid="next-button"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Next
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
